DESIGN.md

```diff
@@ -0,0 +1,247 @@
+
+Our first implementation of this interface will be part of the libary and as simple as possible: a memory store tracking the owner, host pid, stdout, stderr, whether complete, and exit code for any requested worker.
+
+This memory `ProcessStore` implementation will:
+
+* Use `bytes.Buffer` for stdout/stderr, making it particularly easy to process as both an `io.Writer` and `io.Reader`. 
+* Store everything in a map/hashtable, in memory, so it's easy for it to get at and write immediately to across our various parts that need to write to and read from it
+
+Our library contracts are designed as follows:
+
+```golang
+type WorkerGroup struct {
+  Store ProcessStore
+}
+
+// NewWorkerGroup will init and return a new worker group with a provider ProcessStore implementation
+func NewWorkerGroup(store ProcessStore) *WorkerGroup
+
+// Start will initiate a new worker/command on behalf of a user/owner
+func (wg *WorkerGroup) Start(cmd []string, ownerID string) (workerID string, err error)
+
+// Stop will take in a worker ID and the ID of the user requesting the stop, determine if a stop is allowed, and then do it
+func (wg *WorkerGroup) Stop(workerID string, requesterID string) error
+
+// GetStatus is a wrapper around the process store implementation to also manage authorization-related logic, e.g. does requester = owner
+func (wg *WorkerGroup) GetStatus(workerID string, requesterID string) (done bool, exitCode int, err error)
+
+// GetOutput is a wrapper around the process store implementation to also manage authorization-related logic, e.g. does requester = owner
+func (wg *WorkerGroup) GetOutput(workerID string, requesterID string) (stdout io.Reader, stderr io.Reader, err error)
+```
+
+### Details of Linux Namespacing and Cgroup Execution
+
+> :bulb: I don't now how we could use things like `syscall`/`sys/unix` `CLONE_INTO_CGROUP` or `SysProcAttr.CgroupFD`. I've never used them, and usage of them/knowledge of how to use them seems _very_ limited in my, well, also limited experience with how they are supposed to work. So, we'll stick to other paths here.
+
+Namespace isolation while executing our user's command is achieved generally by setting [`Cloneflags`/`Unshareflags`](https://pkg.go.dev/syscall#pkg-constants) for the `SysProcAttr` on a standard [`os/exec.Cmd` object](https://pkg.go.dev/os/exec#Cmd). That's the simple part.
+
+However, because of how Golang doesn't support certain ideas around `fork` natively, we'll need to wrap the user-specified command process in a wrapped execution of our own making to ensure that the namespace isolation and cgroup control actually work. Well do that by doing the following:
+
+* Building a pre-built binary as an artifact that should be installed on a system using this library, e.g. `rockholla-go-linux-ns-cgroup-worker-fork` (we will make sure the build/install process and tooling does this automatically, for example if you're installing the built server binary, you'll get this one as well as a dependency)
```

Comment:
From https://github.com/gravitational/careers/blob/main/challenges/systems/challenge-1.md#dependencies

> The server should also not rely on any shell scripts, external binaries or use containers to execute jobs.

Do you see any way to not really on the external binary ?

Reply:
This is a mistake on my part in interpreting the "external binaries" part of the requirements. I was thinking it was no reliance on external binaries not built as part of this project. I will go back to the drawing board here with your feedback and adjust accordingly asap. Thank you for taking the time on this Marek.